xds/internal/test/e2e/e2e_test.go

```diff
@@ -0,0 +1,257 @@
+		SecLevel:   e2e.SecurityLevelNone,
+	})
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	if err := cp.server.Update(ctx, resources); err != nil {
+		t.Fatalf("failed to update control plane resources: %v", err)
+	}
+
+	st, err := c.clientStats(ctx)
+	if err != nil {
+		t.Fatalf("failed to get client stats: %v", err)
+	}
+	if st.NumFailures != 0 {
+		t.Fatalf("Got %v failures: %+v", st.NumFailures, st)
+	}
+}
+
+// TestAffinity covers the affinity tests with ringhash policy.
+// - client is configured to use ringhash, with 3 backends
+// - all RPCs will hash a specific metadata header
+// - verify that
+//   - all RPCs with the same metadata value are sent to the same backend
+//   - only one backend is Ready
+// - send more RPCs with different metadata values until a new backend is picked, and verify that
+//   - only two backends are in Ready
+func TestAffinity(t *testing.T) {
+	const (
+		testName     = "affinity"
+		backendCount = 3
+		testMDKey    = "xds_md"
+		testMDValue  = "unary_yranu"
+	)
+	cp, c, servers, cleanup := setup(t, testOpts{
+		testName:     testName,
+		backendCount: backendCount,
+		clientFlags:  []string{"--rpc=EmptyCall", fmt.Sprintf("--metadata=EmptyCall:%s:%s", testMDKey, testMDValue)},
+	})
+	defer cleanup()
+
+	resources := e2e.DefaultClientResources(e2e.ResourceParams{
+		DialTarget: testName,
+		NodeID:     cp.nodeID,
+		Host:       "localhost",
+		Port:       serverPort,
+		SecLevel:   e2e.SecurityLevelNone,
+	})
+
+	// Update EDS to multiple backends.
+	var ports []uint32
+	for _, s := range servers {
+		ports = append(ports, uint32(s.port))
+	}
+	edsMsg := resources.Endpoints[0]
+	resources.Endpoints[0] = e2e.DefaultEndpoint(
+		edsMsg.ClusterName,
+		"localhost",
+		ports,
+	)
+
+	// Update CDS lbpolicy to ringhash.
+	cdsMsg := resources.Clusters[0]
+	cdsMsg.LbPolicy = v3clusterpb.Cluster_RING_HASH
+
+	// Update RDS to hash the header.
+	rdsMsg := resources.Routes[0]
+	rdsMsg.VirtualHosts[0].Routes[0].Action = &v3routepb.Route_Route{Route: &v3routepb.RouteAction{
+		ClusterSpecifier: &v3routepb.RouteAction_Cluster{Cluster: cdsMsg.Name},
+		HashPolicy: []*v3routepb.RouteAction_HashPolicy{{
+			PolicySpecifier: &v3routepb.RouteAction_HashPolicy_Header_{
+				Header: &v3routepb.RouteAction_HashPolicy_Header{
+					HeaderName: testMDKey,
+				},
+			},
+		}},
+	}}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	if err := cp.server.Update(ctx, resources); err != nil {
+		t.Fatalf("failed to update control plane resources: %v", err)
+	}
+
+	// We can skip CSDS check because there's no long delay as in TD.
+
+	st, err := c.clientStats(ctx)
+	if err != nil {
+		t.Fatalf("failed to get client stats: %v", err)
+	}
+	if st.NumFailures != 0 {
+		t.Fatalf("Got %v failures: %+v", st.NumFailures, st)
+	}
+	if len(st.RpcsByPeer) != 1 {
+		t.Fatalf("more than 1 backends got traffic: %v, want 1", st.RpcsByPeer)
+	}
```

Comment:
Would some kind of a retry behavior when querying for these client stats be useful? Otherwise we could be prone to updates not reaching fast enough or updates not being applied fast enough. Or have things been super stable in your testing so far that you don't think it is required?

Reply:
For this and the other stats checks in this PR, it's not necessary. There's actually no race.

We only update the control plane once at the init (so there's only one version of the xds resources).
Before that, all the RPCs are waiting for the picker, so they only make progress after EDS is received, and the picker is built.

The other RPC stats check in the affinity test happens after we know a second peer is picked. So that also doesn't race.

In the future, if the test updates xds resources in the middle, we may need to add a retry, or maybe just do a CSDS check.